### PR TITLE
Feature/sdl 0238 keyboard enhancements

### PR DESCRIPTION
**Pull request by AKalinich-Luxoft**
Implements #208 

This PR is **ready** for review.

### Testing Plan
Tested manually

**Note:** [Manual_instruction_Keyboard_Enhancements_v1.2.zip](https://github.com/smartdevicelink/sdl_hmi/files/5824244/Manual_instruction_Keyboard_Enhancements_v1.2.zip)

### Summary
- Add numeric keyboard implementation
- Add keyboard controller with internal logic
- Align keyboard layouts and style with proposal
- Make proper processing of SetGlobalProperties
- Set default app global properties
- Add capabilities for new layouts and properties

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by AKalinich-Luxoft** [comment]
@ashwink11, this PR is ready for review

**Comment by ashwink11** [comment]
@AKalinich-Luxoft changes approved from the Ford side.

**Comment by KhrystynaDubovyk** [comment]
@JackLivio This PR is ready for Livio ready. Thank you!

**Comment by ashwink11** [comment]
@jordynmackool  this PR is ready for review. Thanks!

**Review by jacobkeeler — CHANGES_REQUESTED** [review]
Note that this is a partial review due to the fact that there are some pending revisions which might affect the implementation of this slightly (see https://github.com/smartdevicelink/sdl_evolution/pull/1117)

File: app/controller/sdl/KeyboardController.js [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+    /**
+     * Flag to check should masking button be displayed or not
+     */
+    showMaskButton: false,
+
+    /**
+     * Array of symbols not allowed to show on keyboard UI
+     */
+    unsupportedKeyboardSymbols: [ '^' ],
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply app/controller/sdl/KeyboardController.js:54]: @jacobkeeler it's just a random symbol we decided to chose for our testing purposes and to demonstrate that this part of functionality works as expected

File: app/controller/sdl/KeyboardController.js [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+    isLayoutActive: function(layout) {
+        const default_layout = "QWERTY";
+
+        if (SDL.SDLController.model == null && layout == default_layout) {
+            return true;
+        }
+
+        return SDL.SDLController.model ?
+            SDL.SDLController.model.globalProperties.keyboardProperties.keyboardLayout == layout :
+            false;
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply app/controller/sdl/KeyboardController.js]: @AKalinich-Luxoft fixed in https://github.com/smartdevicelink/sdl_hmi/pull/474/commits/02c7de3fc7457bc4abe5732264b05b5b40f33906

File: app/controller/sdl/KeyboardController.js [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        if (SDL.SDLController.model &&
+          SDL.SDLController.model.activeRequests.uiPerformInteraction) {
+          SDL.SDLController.onResetTimeout(
+            SDL.SDLController.model.appID, 'UI.PerformInteraction'
+          );
+        }
+    },
+
+    /**
+     * @description Disables or enables charaters depeding on global properties
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply app/controller/sdl/KeyboardController.js]: @jacobkeeler fixed in https://github.com/smartdevicelink/sdl_hmi/pull/474/commits/02c7de3fc7457bc4abe5732264b05b5b40f33906

File: app/controller/sdl/KeyboardController.js [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+     * @description Closes keyboard view and cancels active interaction request
+     */
+    closeKeyboardView: function() {
+        if (SDL.SDLController.model &&
+            SDL.SDLController.model.activeRequests.uiPerformInteraction &&
+            !SDL.InteractionChoicesView.active) {
+            FFW.UI.OnKeyboardInput('', 'ENTRY_CANCELLED');
+            SDL.InteractionChoicesView.deactivate('ABORTED');
+          }
+          SDL.Keyboard.deactivate();
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply app/controller/sdl/KeyboardController.js]: @AKalinich-Luxoft fixed in https://github.com/smartdevicelink/sdl_hmi/pull/474/commits/02c7de3fc7457bc4abe5732264b05b5b40f33906

File: app/controller/sdl/KeyboardController.js [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+    /**
+     * @description Updates keyboard input according to current values
+     * of internal controller flags
+     */
+    updateInputMasking: function() {
+      if (SDL.Keyboard) {
+        if (SDL.KeyboardController.maskCharacters) {
+          SDL.Keyboard.searchBar.input.type = 'password';
+        } else {
+            SDL.Keyboard.searchBar.input.type = 'text';
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply app/controller/sdl/KeyboardController.js]: @jacobkeeler fixed in https://github.com/smartdevicelink/sdl_hmi/pull/474/commits/02c7de3fc7457bc4abe5732264b05b5b40f33906

**Comment by jordynmackool** [comment]
@AKalinich-Luxoft the proposal markdown file has been updated per the revisions included in the accepted with revisions review issue: [Revise SDL-0238 Keyboard Enhancements](https://github.com/smartdevicelink/sdl_evolution/pull/1119).

Please make the needed updates to this PR and then tag @jacobkeeler to review when ready. Thanks!

**Comment by AKalinich-Luxoft** [comment]
@jacobkeeler this PR has been updated according to revision and ready for Livio review

File: app/controller/sdl/KeyboardController.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacobkeeler) [reply app/controller/sdl/KeyboardController.js:54]: Alright, but I don't see any reason for this feature to be included in the first place, as it's not mentioned in the proposal at any point

[the remark this answers is not in the document]
Comment (jacobkeeler) [reply app/controller/sdl/KeyboardController.js:54]: That said, considering it's a fairly small feature, there isn't much point in removing it at this point, so I'll mark this as resolved

[the remark this answers is not in the document]
Comment (jacobkeeler) [reply app/controller/sdl/KeyboardController.js:54]: Never mind, I didn't notice it initially, but I now see that this functionality is explicitly listed in the proposal: https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0238-Keyboard-Enhancements.md#change-in-keyboardproperties-struct

**Review by jacobkeeler — APPROVED** [review]
As a note, I did notice that the documentation for this feature says that the HMI should avoid duplicating custom keys, but this implementation does not do so. That said, I would see this as an optional optimization, so I am approving this PR.

**Pull request merged**